### PR TITLE
Add accessor methods to neighbor/child/router objects.

### PR DESCRIPTION
**Pull request by jwhui**
This PR makes no changes to existing logic.

### CODE REVIEW

File: src/core/api/thread_api.cpp [hunk 1]
```diff
@@ -479,17 +479,17 @@ ThreadError otThreadGetParentInfo(otInstance *aInstance, otRouterInfo *aParentIn
...
-    aParentInfo->mRloc16          = parent->mValid.mRloc16;
-    aParentInfo->mRouterId        = Mle::Mle::GetRouterId(parent->mValid.mRloc16);
-    aParentInfo->mNextHop         = parent->mNextHop;
-    aParentInfo->mPathCost        = parent->mCost;
-    aParentInfo->mLinkQualityIn   = parent->mLinkInfo.GetLinkQuality(aInstance->mThreadNetif.GetMac().GetNoiseFloor());
-    aParentInfo->mLinkQualityOut  = parent->mLinkQualityOut;
-    aParentInfo->mAge             = static_cast<uint8_t>(Timer::MsecToSec(Timer::GetNow() - parent->mLastHeard));
-    aParentInfo->mAllocated       = parent->mAllocated;
-    aParentInfo->mLinkEstablished = parent->mState == Neighbor::kStateValid;
+    memcpy(aParentInfo->mExtAddress.m8, &parent->GetExtAddress(), OT_EXT_ADDRESS_SIZE);
```
Comment (turon) [inline src/core/api/thread_api.cpp]: Using OT_EXT_ADDRESS_SIZE here, but sizeof(parent->GetExtAddress()) elsewhere.
Should the code follow exactly one pattern?

File: src/core/mac/mac.cpp [hunk 2]
```diff
@@ -1316,7 +1316,7 @@ void Mac::ReceiveDoneTask(Frame *aFrame, ThreadError aError)
         }
 
         srcaddr.mLength = sizeof(srcaddr.mExtAddress);
-        memcpy(&srcaddr.mExtAddress, &neighbor->mMacAddr, sizeof(srcaddr.mExtAddress));
+        srcaddr.mExtAddress = neighbor->GetExtAddress();
```
Comment (turon) [inline src/core/mac/mac.cpp]: :+1: Kind of cool that c++ pass-by-reference assignment just does the copy for you.

File: src/core/thread/address_resolver.cpp [hunk 3]
```diff
@@ -529,24 +529,22 @@ void AddressResolver::HandleAddressQuery(Coap::Header &aHeader, Message &aMessag
...
+            children[i].GetLinkFailures() >= Mle::kFailedChildTransmissions)
         {
             continue;
         }
 
-        for (int j = 0; j < Child::kMaxIp6AddressPerChild; j++)
+        for (uint8_t j = 0; j < Child::kMaxIp6AddressPerChild; j++)
         {
-            if (memcmp(&children[i].mIp6Address[j], targetTlv.GetTarget(), sizeof(children[i].mIp6Address[j])))
+            if (children[i].GetIp6Address(j) != *targetTlv.GetTarget())
```
Comment (turon) [inline src/core/thread/address_resolver.cpp]: :+1: Another nice simplification of the compare automatically handling the 64-bit values.

File: src/core/thread/topology.hpp [hunk 4]
```diff
@@ -50,33 +51,11 @@ namespace Thread {
...
-
-    uint32_t mKeySequence;               ///< Current key sequence
-
     /**
      * Neighbor link states.
      *
      */
     enum State
-#if _WIN32
-        : unsigned int
```
Comment (nibanks) [inline src/core/thread/topology.hpp]: Why did you remove this? Have you verified that Windows still can successfully connect?

File: src/core/thread/mesh_forwarder.cpp [hunk 5]
```diff
@@ -911,8 +910,8 @@ ThreadError MeshForwarder::HandleFrameRequest(Mac::Frame &aFrame)
     // already checked and handled in `SendFragment()` method.
 
     if (((child = mNetif.GetMle().GetChild(macDest)) != NULL)
-        && ((child->mMode & Mle::ModeTlv::kModeRxOnWhenIdle) == 0)
-        && (child->GetQueuedMessageCount() > 1))
+        && !child->GetRxOnWhenIdle()
```
Comment (turon) [inline src/core/thread/mesh_forwarder.cpp]: nit: I know `RxOnWhenIdle` terminology follows the spec strictly, and originates from IEEE 802.15.4-2006 of old, but I feel like `child->IsSleepy()` would be more readable.

File: src/core/thread/topology.hpp (hunk 4 above)
Comment (jwhui) [inline src/core/thread/topology.hpp]: Few other related changes:
1. `mState` type was changed to `uint8_t`.
2. `static_cast<>` is used to convert between the `uint8_t` and `State` types.
3. `test_toolchain.cpp` is also updated to validate the `GetState()`/`SetState()` methods.

I would hope these changes avoid the need for the `if _WIN32` pre-processor statement.

File: src/core/thread/mesh_forwarder.cpp [hunk 6]
```diff
@@ -1957,12 +1963,12 @@ void MeshForwarder::HandleDataRequest(const Mac::Address &aMacSource, const Thre
...
-    child->mLastHeard = Timer::GetNow();
-    child->mLinkFailures = 0;
+    child->SetLastHeard(Timer::GetNow());
+    child->ResetLinkFailures();
 
-    if (!mSourceMatchController.IsEnabled() || (child->GetQueuedMessageCount() > 0))
+    if (!mSourceMatchController.IsEnabled() || (child->GetIndirectMessageCount() > 0))
     {
-        child->mDataRequest = true;
+        child->SetDataRequestPending(true);
```
Comment (turon) [inline src/core/thread/mesh_forwarder.cpp]: nit: Could `child->SetFramePending(true)` be a shorter, more descriptive name that aligns with the actual name of the frame control bit set by the parent?

File: src/core/thread/mle_router.cpp [hunk 7]
```diff
@@ -747,15 +743,15 @@ ThreadError MleRouter::SendLinkAccept(const Ip6::MessageInfo &aMessageInfo, Neig
     linkMargin = LinkQualityInfo::ConvertRssToLinkMargin(mNetif.GetMac().GetNoiseFloor(), threadMessageInfo->mRss);
 
     // add for certification testing
-    if (isAssignLinkQuality &&
-        (memcmp(aNeighbor->mMacAddr.m8, mAddr64.m8, OT_EXT_ADDRESS_SIZE) == 0))
+    if (isAssignLinkQuality && aNeighbor != NULL &&
+        (memcmp(&aNeighbor->GetExtAddress(), mAddr64.m8, OT_EXT_ADDRESS_SIZE) == 0))
```
Comment (turon) [inline src/core/thread/mle_router.cpp]: Would the `==` operator work here as it did earlier?

File: src/core/thread/mle_router.cpp [hunk 8]
```diff
@@ -1359,11 +1351,11 @@ ThreadError MleRouter::HandleAdvertisement(const Message &aMessage, const Ip6::M
             ExitNow();
         }
 
-        if (memcmp(&mParent.mMacAddr, &macAddr, sizeof(mParent.mMacAddr)) == 0)
+        if (memcmp(&mParent.GetExtAddress(), &macAddr, sizeof(mParent.GetExtAddress())) == 0)
```
Comment (turon) [inline src/core/thread/mle_router.cpp]: Switch to c++ `==` operator as done earlier?

File: src/core/thread/mle_router.cpp [hunk 9]
```diff
@@ -2056,64 +2045,67 @@ ThreadError MleRouter::HandleChildIdRequest(const Message &aMessage, const Ip6::
     // Remove from router table
     for (int i = 0; i <= kMaxRouterId; i++)
     {
-        if (mRouters[i].mState != Neighbor::kStateInvalid &&
-            memcmp(&mRouters[i].mMacAddr, &macAddr, sizeof(mRouters[i].mMacAddr)) == 0)
+        if (mRouters[i].GetState() != Neighbor::kStateInvalid &&
+            memcmp(&mRouters[i].GetExtAddress(), &macAddr, sizeof(macAddr)) == 0)
```
Comment (turon) [inline src/core/thread/mle_router.cpp]: c++ `==` operator?

File: src/core/thread/mle_router.cpp [hunk 10]
```diff
@@ -2620,17 +2612,17 @@ ThreadError MleRouter::SendChildIdResponse(Child *aChild)
...
 
-    for (uint8_t i = 0; i < sizeof(aChild->mRequestTlvs); i++)
+    for (uint8_t i = 0; i < Child::kMaxRequestTlvs; i++)
     {
-        switch (aChild->mRequestTlvs[i])
+        switch (aChild->GetRequestTlv(i))
         {
         case Tlv::kNetworkData:
-            SuccessOrExit(error = AppendNetworkData(*message, (aChild->mMode & ModeTlv::kModeFullNetworkData) == 0));
+            SuccessOrExit(error = AppendNetworkData(*message, !aChild->GetFullNetworkData()));
```
Comment (turon) [inline src/core/thread/mle_router.cpp]: Clarity check: a naive reading of `GetFullNetworkData` would be that the method returns a large blob of data, not a boolean.  Would `HasFullNetworkData` be better?

File: src/core/thread/mle_router.cpp [hunk 11]
```diff
@@ -2650,21 +2642,21 @@ ThreadError MleRouter::SendChildIdResponse(Child *aChild)
         }
     }
 
-    if ((aChild->mMode & ModeTlv::kModeFFD) == 0)
+    if (!aChild->GetFullThreadDevice())
```
Comment (turon) [inline src/core/thread/mle_router.cpp]: Likewise, `IsFullThreadDevice` may be a better method name for this boolean.

File: src/core/thread/mesh_forwarder.cpp (hunk 6 above)
Comment (jwhui) [inline src/core/thread/mesh_forwarder.cpp]: `SetDataRequestPending()` is to record the fact that an IEEE 802.15.4 Data Request message was received from the child and needs to be serviced.

File: src/core/thread/mle_router.cpp (hunk 7 above)
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: That operator is not yet defined, was planning to do that general cleanup a separate PR.

File: src/core/thread/mle_router.cpp (hunk 8 above)
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: That operator is not yet defined, was planning to do that general cleanup a separate PR.

File: src/core/thread/mle_router.cpp (hunk 9 above)
Comment (jwhui) [inline src/core/thread/mle_router.cpp]: That operator is not yet defined, was planning to do that general cleanup a separate PR.

File: src/core/thread/src_match_controller.cpp [hunk 12]
```diff
@@ -174,25 +178,25 @@ void SourceMatchController::ClearEntry(Child &aChild)
...
+        error = otPlatRadioClearSrcMatchShortEntry(GetInstance(), aChild.GetRloc16());
 
         otLogDebgMac(GetInstance(), "SrcAddrMatch - Clearing short address: 0x%04x -- %s (%d)",
-                     aChild.mValid.mRloc16, otThreadErrorToString(error), error);
+                     aChild.GetRloc16(), otThreadErrorToString(error), error);
     }
     else
     {
-        uint8_t addr[sizeof(aChild.mMacAddr)];
+        uint8_t addr[sizeof(aChild.GetExtAddress())];
```
Comment (abtink) [inline src/core/thread/src_match_controller.cpp]: I think it maybe safer to not use the `sizeof(aChild.GetExtAddress())` here and use somethig like `sizeof(Mac::ExtAddress)` instead.

We are getting `sizeof` of return value of a function that returns a reference to `Mac::ExtAddress` (`const Mac::ExtAddress &GetExtAddress()`)... I am not sure if `sizeof` gives the size of the `Mac::ExtAddress` or sizeof a reference to it (which will then be incorrect).

File: src/core/thread/topology.hpp [hunk 13]
```diff
@@ -129,16 +370,275 @@ class Child : public Neighbor
...
+     */
+    uint8_t GetChallengeSize(void) const { return sizeof(mAttachChallenge); }
+
+    /**
+     * This method gets the IEEE 802.15.4 Frame Counter used during indirect retransmissions.
+     *
+     * @returns The IEEE 802.15.4 Frame Counter value.
+     *
+     */
+    uint32_t GetIndirectFrameCounter(void) const { return mIndirectSendInfo.mFrameCounter; }
```
Comment (abtink) [inline src/core/thread/topology.hpp]: We can consider getting rid of `mIndirectSendInfo`struct as well... Mainly used this to group all indirect info together... (I can do it in a operate PR later)...

File: src/core/thread/src_match_controller.cpp (hunk 12 above)
Comment (jwhui) [inline src/core/thread/src_match_controller.cpp]: From the [C++ standard](http://en.cppreference.com/w/cpp/language/sizeof):

```
When applied to a reference type, the result is the size of the referenced type.
```

Comment (abtink) [inline src/core/thread/src_match_controller.cpp]: Good to know. Thanks.

File: src/core/thread/topology.hpp (hunk 4 above)
Comment (jwhui) [inline src/core/thread/topology.hpp]: @nibanks, are you OK with these changes?

Comment (nibanks) [inline src/core/thread/topology.hpp]: Yes. It looks like it would be OK, though I haven't actually validated it.
